server/src/main/java/org/candlepin/resteasy/filter/CandlepinQueryInterceptor.java

```diff
@@ -55,101 +44,118 @@
-                    pageRequest.getSortBy() :
-                    AbstractHibernateObject.DEFAULT_SORT_FIELD;
-
-                PageRequest.Order order = pageRequest.getOrder() != null ?
-                    pageRequest.getOrder() :
-                    PageRequest.DEFAULT_ORDER;
-
-                query.addOrder(order == PageRequest.Order.DESCENDING ?
-                    Order.desc(sortField) :
-                    Order.asc(sortField)
-                );
-
-                if (pageRequest.isPaging()) {
-                    query.setFirstResult((pageRequest.getPage() - 1) * pageRequest.getPerPage());
-                    query.setMaxResults(pageRequest.getPerPage());
-
-                    // Create a page object for the link header response
-                    Page page = new Page();
-                    page.setMaxRecords(query.getRowCount()); // This is expensive :(
-                    page.setPageRequest(pageRequest);
-                    // Note: we don't need to store the page data in the page
-
-                    ResteasyProviderFactory.pushContext(Page.class, page);
+            try {
+                final CandlepinQuery query = (CandlepinQuery) entity;
+
+                // Use a separate session so we aren't at risk of lazy loading or interceptors closing
+                // our cursor mid-stream.
+                query.useSession(session);
+
+                // Apply any paging config we may have
+                applyPaging(pageRequest, query);
+
+                // Set the output streamer that will stream our query result
+                response.setEntity(buildOutputStreamer(session, query));
+            }
+            catch (final RuntimeException e) {
+                if (session != null) {
+                    session.close();
                 }
+                throw e;
             }
+        }
+    }
 
-            // Set the output streamer that will stream our query result
-            response.setEntity(new StreamingOutput() {
-                @Override
-                public void write(OutputStream stream) throws IOException, WebApplicationException {
-                    try (JsonGenerator generator = mapper.getJsonFactory().createGenerator(stream);
-                        ResultIterator<Object> iterator = query.iterate()) {
-
-                        generator.writeStartArray();
-
-                        while (iterator.hasNext()) {
-                            mapper.writeValue(generator, iterator.next());
-                        }
-
-                        generator.writeEndArray();
-                        generator.flush();
-                    }
-                    finally {
-                        if (session != null) {
-                            session.close();
-                        }
-                    }
-                }
-            });
+    private void applyPaging(final PageRequest pageRequest, final CandlepinQuery query) {
```

Comment:
Just clarity. Method is quite long so I extracted its nested responsibilities to name them.

Reply:
That's what the comments are for. Keep in mind that for a human reader, you're adding new context jumps for them to keep up with.

Reply:
I am sorry but I don't agree with you. Tautological comments are a code smell since they are just duplicating what the code already says. Code is for explaining what you are doing. Comments are for explaining why. While I agree that mental inlining and context jumps can be a problem, mental grouping and context overload is the same problem just opposite side of it.

Reply:
I believe this is simply a case of different styles of code. Reducing the cognitive load by factoring out methods is fine with me, especially in this case where the original method was creating an anonymous class and passing it elsewhere at once.